File: android/src/main/kotlin/com/klarna/inapp/sdk/flutter_klarna_inapp_sdk/webview/WebViewMethod.kt
package com.klarna.inapp.sdk.flutter_klarna_inapp_sdk.webview

import com.klarna.inapp.sdk.flutter_klarna_inapp_sdk.core.method.MethodParser
import com.klarna.inapp.sdk.flutter_klarna_inapp_sdk.core.util.requireArgument
import io.flutter.plugin.common.MethodCall

internal sealed class WebViewMethod {
    object Initialize : WebViewMethod()
    object Destroy : WebViewMethod()
    object Show : WebViewMethod()
    object Hide : WebViewMethod()
    class LoadURL(val url: String) : WebViewMethod()
    class LoadJS(val js: String) : WebViewMethod()

    internal object Parser : MethodParser<WebViewMethod> {
        override fun parse(call: MethodCall): WebViewMethod? {
            return when (call.method) {
                "initialize" -> Initialize
                "destroy" -> Destroy
                "show" -> Show
                "hide" -> Hide
                "loadURL" -> LoadURL(call.requireArgument("url"))
                "loadJS" -> LoadJS(call.requireArgument("js"))
                else -> null
            }
        }
    }
}

File: ios/Classes/WebView/WebViewMethod.swift
import Flutter

protocol WebViewMethod {
    
}

class WebViewMethods {
    
    struct Initialize: WebViewMethod {}
    struct Destroy: WebViewMethod {}
    struct Show: WebViewMethod {}
    struct Hide: WebViewMethod {}
    struct LoadURL: WebViewMethod {
        let url: String
    }
    struct LoadJS: WebViewMethod {
        let js: String
    }
    struct AddToHybridSDK: WebViewMethod {}

    class Parser: MethodParser<WebViewMethod> {
        override func parse(call: FlutterMethodCall) throws -> WebViewMethod? {
            switch call.method {
            case "initialize":
                return Initialize()
            case "destroy":
                return Destroy()
            case "show":
                return Show()
            case "hide":
                return Hide()
            case "loadURL":
                return LoadURL(url: try call.requireArgument(key: "url"))
            case "loadJS":
                return LoadJS(js: try call.requireArgument(key: "js"))
            case "addToHybridSdk":
                return AddToHybridSDK()
            default:
                return nil
            }
        }
    }
}
